dataproc/single_job_workflow.py

```diff
@@ -0,0 +1,215 @@
+    workflow_template_service_grpc_transport,
+)
+
+DEFAULT_FILENAME = "pyspark_sort.py"
+waiting_callback = False
+
+
+def get_default_pyspark_file():
+    """Gets the PySpark file from this directory."""
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    f = open(os.path.join(current_dir, DEFAULT_FILENAME), "rb")
+    return f, DEFAULT_FILENAME
+
+
+def get_pyspark_file(filename):
+    f = open(filename, "rb")
+    return f, os.path.basename(filename)
+
+
+def get_region_from_zone(zone):
+    try:
+        region_as_list = zone.split("-")[:-1]
+        return "-".join(region_as_list)
+    except (AttributeError, IndexError, ValueError):
+        raise ValueError("Invalid zone provided, please check your input.")
+
+
+def upload_pyspark_file(project, bucket_name, filename, spark_file):
+    """Uploads the PySpark file in this directory to the configured input
+    bucket."""
+    print("Uploading pyspark file to Cloud Storage.")
+    client = storage.Client(project=project)
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(filename)
+    blob.upload_from_file(spark_file)
+
+
+def run_workflow(dataproc, project, region, zone, bucket_name, filename,
+                 cluster_name):
+    print("filename = {} ".format(filename))
+    parent = "projects/{}/regions/{}".format(project, region)
+    zone_uri = ("https://www.googleapis.com/compute/v1/projects/{}/zones/{}"
+                .format(project, zone))
+
+    workflow_data = {
+        "placement": {
+            "managed_cluster": {
+                "cluster_name": cluster_name,
+                "config": {
+                    "gce_cluster_config": {"zone_uri": zone_uri},
+                    "master_config": {
+                        "num_instances": 1,
+                        "machine_type_uri": "n1-standard-1",
+                    },
+                    "worker_config": {
+                        "num_instances": 2,
+                        "machine_type_uri": "n1-standard-1",
+                    },
+                },
+            }
+        },
+        "jobs": [
+            {
+                "pyspark_job": {
+                    "main_python_file_uri": "gs://{}/{}".format(
+                        bucket_name, filename)
+                },
+                "step_id": "pyspark-job",
+            }
+        ],
+    }
+
+    workflow = dataproc.instantiate_inline_workflow_template(parent,
+                                                             workflow_data)
+
+    workflow.add_done_callback(callback)
+    global waiting_callback
+    waiting_callback = True
+
+
+def callback(operation_future):
+    # Reset global when callback returns.
+    global waiting_callback
+    waiting_callback = False
+
+
+def wait_for_workflow_end():
+    """Wait for cluster creation."""
+    print("Waiting for workflow completion ...")
+    print('Workflow and job progress, and job driver output available from: '
```

Comment:
Don't mix double-quotes and single quotes for strings in the same file. The only exception is when a string contains a double quote.

I actually recommend running `black` https://black.readthedocs.io/en/stable/ to fix issues like this (but then you'll probably have to clean up some of the line lengths since black doesn't respect the 79 character limit.)

Reply:
Done.